### PR TITLE
use serial interface for keypad

### PR DESCRIPTION
**Pull request by littleaj**
Using a 4x4 multiplexing keypad with serial interface. 74hc595 drives HIGHs to columns, 74hc165 reads rows.

### CODE REVIEW

**Comment by littleaj** [comment]
@dctucker code was tested and it works. just looking for optimization opportunities

File: src/temp-alarm.ino [hunk 1]
```diff
@@ -75,24 +102,97 @@ unsigned long elapsed() {
...
+void write595(uint8_t b) {
+    digitalWrite(PIN_SS, LOW); // enable 595
+    asm("nop\n nop\n"); // short delay
+    SPI.transfer(b); // send byte
+    digitalWrite(PIN_SS, HIGH); // disable 595
+}
+
+uint8_t read165() {
+    digitalWrite(PIN_LOAD, LOW); // pulse load
+    asm("nop\n nop\n");
```
Comment (dctucker) [inline src/temp-alarm.ino:163]: Is this a clock-cycle buffer with specific timing? A comment might be useful.

File: src/temp-alarm.ino [hunk 2]
```diff
@@ -1,40 +1,45 @@
...
 
 const unsigned long startTime = millis();
 
-const char hexaKeys[KEYPAD_ROWS][KEYPAD_COLS] = {
-    { '1', '2', '3', 'A' },
-    { '4', '5', '6', 'B' },
-    { '7', '8', '9', 'C' },
-    { '*', '0', '#', 'D' },
+// TODO I could swap row/column wires to make this array "right-side up"
+const char hexaKeys[KEYPAD_ROWS*KEYPAD_COLS] = {
```
Comment (dctucker) [inline src/temp-alarm.ino:31]: Is there efficiency to be gained by switching to a one-dimensional array, or is it mostly ａｅｓｔｈｅｔｉｃ?

File: src/temp-alarm.ino [hunk 3]
```diff
@@ -75,24 +102,97 @@ unsigned long elapsed() {
...
+    if (currentKey) {
+        lastKey = currentKey;
+    } else {
+        lastKey = 0;
+    }
+    return currentKey;
+}
+
+uint8_t readKey() {
+    // TODO
```
Comment (dctucker) [inline src/temp-alarm.ino]: what specifically?

**Review by dctucker — APPROVED** [review]
Looks good, just some minor nits

File: src/temp-alarm.ino (hunk 1 above)
Comment (littleaj) [reply src/temp-alarm.ino:163]: just needs to be a short delay. I saw someone do this instead of using `delay`

File: src/temp-alarm.ino (hunk 2 above)
Comment (littleaj) [reply src/temp-alarm.ino:31]: yeah, the key read is the index into this array. I mixed up the wiring which is why I had to transpose it.

File: src/temp-alarm.ino (hunk 3 above)
Comment (littleaj) [reply src/temp-alarm.ino]: it's done :)

**Pull request merged**